sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala

```diff
@@ -344,4 +346,13 @@ object Mask extends SupportsNamedArguments {
       strArg, upperCharArg, lowerCharArg, digitCharArg, otherCharArg))
     Seq(functionSignature)
   }
+
+  override def build(funcName: String, expressions: Seq[Expression]): Expression = {
+    if (expressions.length < 1 || expressions.length > 5) {
+      throw QueryCompilationErrors.wrongNumArgsError(
+        funcName, Seq(1, 2, 3, 4, 5), expressions.length)
+    }
+    new Mask(expressions(0), expressions(1), expressions(2), expressions(3), expressions(4))
```

Comment:
What happen when there is no named argument but users provide only 3 arguments?

Reply:
We can't affect the current usage without named arguments. e.g. we can't reject these cases. It will become a breaking change.

Reply:
The expression builder way shrifts the responsibility to find a 'right' constructor (constructor with the right number of arguments) from the `FunctionRegistryBase.build` that uses reflection, to the expression builder itself. I'm not sure if this is a preferred way @cloud-fan  - to really make this `mask` function work, we will need to write argument number matching in the `build` function of the expression builder, e.g. `if (expressions.size == 2) .. `.

Reply:
@anchovYu Thanks for the comments and the thoughts! About this: you are right in that if we want to avoid reflection entirely, we will not be using the ```FunctionRegistryBase.build``` method since that heavily utilizes reflection to find the constructor we want for building the function expression. Therefore, it is true that in cases where we only provide positional arguments that the old version of ```rearrange``` did not handle this case. 

After some thoughts however, I realized that we can implement default arguments as well in a cleaner fashion that what we had with the old way where we would have multiple constructors defined for a single function expression. The ```rearrange``` method would now also be responsible for filling in default values for a function invocation where only positional arguments are provided and if some optional parameter values are left unspecified. In another comment, you were wondering why I checked if the index was -1 when splitting the argument list into positional and named arguments, that check was done precisely to allow the processing of default values for parameters in the case where only positional arguments are provided. I thought this was a fairly elegant way of doing it, although it can be quite confusing. I will definitely need to clean up the code/add some comments to make this more clear.

Reply:
Have discussed with multiple people, I vote for "also be responsible for filling in default values for a function invocation where only positional arguments are provided and if some optional parameter values are left unspecified". We will enforce the required parameters defined before optional ones.

Reply:
Yes, +1 on this approach as well.

Eventually, we can considering putting more required properties in the function signatures, such that most function expression classes do not have to implement any matching logic between provided vs. required arguments at all, and it will be handled generically by the analyzer instead. For now, keeping just the argument names and optional default values in the function signature lets us accomplish the goal of supporting named arguments without breaking any existing use cases.